### PR TITLE
[Mercure] Compatibility with the Docker integration and various improvements

### PR DESCRIPTION
**Pull request by dunglas**
Includes https://github.com/symfony/symfony-docs/pull/16151. Closes dunglas/symfony-docker#200.

### CODE REVIEW

File: mercure.rst [hunk 1]
```diff
@@ -95,57 +90,70 @@ The preferred way to configure MercureBundle is using
...
+                    'jwt' => [
+                        'secret' => '!ChangeMe!',
+                    ],
+                ],
+            ],
+        ]);
 
-If you don't want to use the provided environment variables,
-use the following configuration:
+Alternatively, it's also possible to pass directly the JWT to use:
```
Comment (azjezz) [inline mercure.rst]: should we recommend using secret over jwt value as it allows managing subscribe and publish topics easier from symfony?

File: mercure.rst [hunk 2]
```diff
@@ -642,14 +679,14 @@ During functional testing you can instead decorate the Hub::
...
 
 .. code-block:: yaml
 
     # config/services_test.yaml
-    App\Tests\Functional\Fixtures\HubStub:
-        decorates: mercure.hub.default
+    mercure.hub.default:
+        class: App\Tests\Functional\Stub\HubStub
 
 .. tip::
```
Comment (azjezz) [inline mercure.rst]: since you can have multiple hubs in symfony, we should mention that this will only replace the default one.

**Review by azjezz — COMMENTED** [review]
I think the JWT section could be extended to mention the `jwt.publish`, `jwt.publish`, `jwt.algorithm`, and using custom JWT provider/factory.

File: mercure.rst [hunk 3]
```diff
@@ -624,10 +661,10 @@ You can instead make use of the `MockHub`::
         }
     }
 
-During functional testing you can instead decorate the Hub::
+During functional testing you can instead create a stub of the Hub::
```
Comment (chalasr) [inline mercure.rst]:
```suggestion
For functional testing, you can instead create a stub of the Hub::
```

File: mercure.rst [hunk 4]
```diff
@@ -333,9 +362,9 @@ by using the ``AbstractController::addLink`` helper method::
 
     class DiscoverController extends AbstractController
     {
-        public function __invoke(Request $request, Discovery $discovery): JsonResponse
+        public function discover(Request $request, Discovery $discovery): JsonResponse
         {
-            // Link: <http://localhost:3000/.well-known/mercure>; rel="mercure"
+            // Link: <http://hub.example.com/.well-known/mercure>; rel="mercure"
```
Comment (chalasr) [inline mercure.rst]: Which of `http` and `https` is the right scheme using the flex integration? There's a mix both in the document, can we stick to one and maybe mention that the hub should be served over TLS in prod?

File: mercure.rst (hunk 1 above)
Comment (dunglas) [reply mercure.rst]: It depends of the use case, I changed this section.

File: mercure.rst (hunk 4 above)
Comment (dunglas) [reply mercure.rst]: By the spec, HTTPS is mandatory (in prod at least). The Docker integration doesn't support HTTPS out of the box yet, but I hope that we'll be able to fix that at some point. I updated all the URLs to use HTTPS, as it should be in prod.

**Comment by dunglas** [comment]
@azjezz

> I think the JWT section could be extended to mention the jwt.publish, jwt.publish, jwt.algorithm, and using custom JWT provider/factory.

Indeed, done.

File: mercure.rst [hunk 5]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
+* ``publish``: a list of topics to allow publishing to when generating the JWT
+* ``subscribe``: a list of topics to allow subscribing to when generating the JWT
+* ``algorithm``: The algorithm to use to sign the JWT
+* ``provider``: The ID of a service to call to provide the JWT
+* ``factory``: The ID of a service to call to create the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``factory``: The ID of a service to call to create the JWT (all other options, beside `subscribe`, and `publish` will be ignored)
```

File: mercure.rst [hunk 6]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
+* ``publish``: a list of topics to allow publishing to when generating the JWT
+* ``subscribe``: a list of topics to allow subscribing to when generating the JWT
+* ``algorithm``: The algorithm to use to sign the JWT
+* ``provider``: The ID of a service to call to provide the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``provider``: The ID of a service to call to provide the JWT (all other options will be ignored)
```

File: mercure.rst [hunk 7]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-    This environment variable must contain a JWT, signed with the secret key.
-
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
+* ``publish``: a list of topics to allow publishing to when generating the JWT
+* ``subscribe``: a list of topics to allow subscribing to when generating the JWT
+* ``algorithm``: The algorithm to use to sign the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``algorithm``: The algorithm to use to sign the JWT (only usable when `secret` is provided )
```

File: mercure.rst [hunk 8]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-    Don't put the secret key in ``MERCURE_JWT_SECRET``, it will not work!
-    This environment variable must contain a JWT, signed with the secret key.
-
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
+* ``publish``: a list of topics to allow publishing to when generating the JWT
+* ``subscribe``: a list of topics to allow subscribing to when generating the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``subscribe``: a list of topics to allow subscribing to when generating the JWT (only usable when `secret`, or `factory` are provided )
```

File: mercure.rst [hunk 9]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-
-    Don't put the secret key in ``MERCURE_JWT_SECRET``, it will not work!
-    This environment variable must contain a JWT, signed with the secret key.
-
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
+* ``publish``: a list of topics to allow publishing to when generating the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``publish``: a list of topics to allow publishing to when generating the JWT (only usable when `secret`, or `factory` are provided )
```

File: mercure.rst [hunk 10]
```diff
@@ -95,57 +90,37 @@ The preferred way to configure MercureBundle is using
...
-.. caution::
-
-    Don't put the secret key in ``MERCURE_JWT_SECRET``, it will not work!
-    This environment variable must contain a JWT, signed with the secret key.
-
-    Also, be sure to keep both the secret key and the JWTs... secrets!
-
-If you don't want to use the provided environment variables,
-use the following configuration:
+* ``secret``: the key to use to sign the JWT
```
Comment (azjezz) [inline mercure.rst]:
```suggestion
* ``secret``: the key to use to sign the JWT (all other options, beside `algorithm`, `subscribe`, and `publish` will be ignored)
```

File: mercure.rst [hunk 11]
```diff
@@ -125,6 +133,12 @@ use the following configuration:
                     url: https://mercure-hub.example.com/.well-known/mercure
                     jwt:
                         secret: '!ChangeMe!'
+                        publish: ['foo', 'https://example.com/foo']
+                        subscribe: ['bar', 'https://example.com/bar']
+                        algorithm: 'hmac.sha256'
+                        provider: 'My\Provider'
+                        factory: 'My\Factory'
+                        value: 'my.jwt'
```
Comment (azjezz) [inline mercure.rst:141]: these will be ignored in this case.

File: mercure.rst [hunk 12]
```diff
@@ -135,7 +149,18 @@ use the following configuration:
                 name="default"
                 url="https://mercure-hub.example.com/.well-known/mercure"
             >
-                <jwt secret="!ChangeMe!"/>
+                <jwt
+                    secret="!ChangeMe!"
+                    algorithm="hmac.sha256"
+                    provider="My\Provider"
+                    factory="My\Factory"
+                    value="my.jwt"
```
Comment (azjezz) [inline mercure.rst:157]: these will be ignored in this case.

File: mercure.rst (hunk 11 above)
Comment (dunglas) [reply mercure.rst:141]: Yes it's intended, it's just to document all available keys.

**Review by azjezz — APPROVED** [review]
LGTM

**Comment by javiereguiluz** [comment]
Thanks a lot Kévin

**Pull request merged**